src/manage_sql_agents.c

```diff
@@ -0,0 +1,673 @@
+
+  gchar *insert_agent_uuid = sql_insert (agent_uuid);
+  scanner = sql_int ("SELECT scanner FROM agents WHERE uuid = %s;", insert_agent_uuid);
+  g_free (insert_agent_uuid);
+
+  if (scanner <= 0)
+    {
+      g_warning ("%s: Failed to find scanner for agent UUID %s", __func__, agent_uuid);
+      manage_option_cleanup ();
+      return -1;
+    }
+
+  return scanner;
+}
+
+/**
+ * @brief Synchronize agent data list into the SQL database.
+ *
+ * Performs UPSERT logic: existing agents are updated, new agents
+ * and all IPs are inserted via COPY.
+ *
+ * @param agent_list List of agents to sync.
+ * @return 0 on success, -1 on failure.
+ */
+int
+sync_agents_from_data_list (agent_data_list_t agent_list)
+{
+  if (!agent_list || agent_list->count == 0)
+    return 0;
+
+  db_copy_buffer_t agent_buffer = { 0 };
+  db_copy_buffer_t ip_buffer = { 0 };
+  int status = 0;
+
+  db_copy_buffer_init (
+    &agent_buffer,
+    64 * 1024,
+    "COPY agents (uuid, name, agent_id, hostname, authorized, min_interval, heartbeat_interval,"
+    " connection_status, last_update, schedule, owner, comment, creation_time,"
+    " modification_time, scanner) FROM STDIN;"
+  );
+
+
+  db_copy_buffer_init (
+    &ip_buffer,
+    32 * 1024,
+    "COPY agent_ip_addresses (agent_id, ip_address) FROM STDIN;"
+  );
+
+  sql_begin_immediate ();
+
+  for (int i = 0; i < agent_list->count; ++i)
+    {
+      agent_data_t agent = agent_list->agents[i];
+
+      gboolean exists = agent_column_exists ("agent_id", agent->agent_id);
+
+      if (exists)
+        {
+          update_existing_agent(agent);
+          delete_existing_agent_ips(agent->agent_id);
+        }
+      else
+        {
+          append_agent_row_to_buffer (&agent_buffer, agent);
+        }
+
+      append_ip_rows_to_buffer (&ip_buffer, agent->agent_id, agent->ip_addresses);
+    }
+
+  if (db_copy_buffer_commit (&agent_buffer, TRUE))
+    {
+      g_warning ("%s: COPY for agents failed", __func__);
+      status = -1;
+    }
+
+  if (status == 0)
+    {
+      if (db_copy_buffer_commit (&ip_buffer, TRUE))
+        {
+          g_warning ("%s: COPY for agent_ip_addresses failed", __func__);
+          status = -1;
+        }
+    }
+
+  if (status != 0)
+    sql_rollback ();
+  else
+    sql_commit ();
+
+  db_copy_buffer_cleanup (&agent_buffer);
+  db_copy_buffer_cleanup (&ip_buffer);
+
+  return status;
+}
+
+/**
+ * @brief Initialize SQL-based agent iterator with filtering support.
+ *
+ * @param iterator Pointer to iterator to initialize.
+ * @param get      Get parameters to construct SQL WHERE clause.
+ * @return 0 on success, -1 on failure.
+ */
+int
+init_agent_iterator (iterator_t *iterator, get_data_t *get)
+{
+  g_return_val_if_fail (iterator, -1);
+  g_return_val_if_fail (get, -1);
+
+  static column_t columns[] = AGENT_ITERATOR_COLUMNS;
+  static const char *filter_columns[] = AGENT_ITERATOR_FILTER_COLUMNS;
+
+  g_autofree gchar *id_clause = NULL;
+  g_autofree gchar *full_clause = NULL;
```

Comment:
Using `g_autofree` looks like a good way to simplify temporary strings like that but we'll have to increase the minimum GLib version for this and also consider that this only works with gcc and clang.